src/app/codegen-data-model-provider/EmberDataBuffer.cpp

```diff
@@ -0,0 +1,358 @@
+    case ZCL_INT24U_ATTRIBUTE_TYPE: // Unsigned 24-bit integer
+        return UnsignedDecodeInfo{ 3, 0xFFFFFF };
+    case ZCL_INT32U_ATTRIBUTE_TYPE: // Unsigned 32-bit integer
+        return UnsignedDecodeInfo{ 4, 0xFFFFFFFF };
+    case ZCL_INT40U_ATTRIBUTE_TYPE: // Unsigned 40-bit integer
+        return UnsignedDecodeInfo{ 5, 0xFFFFFFFFFF };
+    case ZCL_INT48U_ATTRIBUTE_TYPE: // Unsigned 48-bit integer
+        return UnsignedDecodeInfo{ 6, 0xFFFFFFFFFFFF };
+    case ZCL_INT56U_ATTRIBUTE_TYPE: // Unsigned 56-bit integer
+        return UnsignedDecodeInfo{ 7, 0xFFFFFFFFFFFFFF };
+    case ZCL_INT64U_ATTRIBUTE_TYPE: // Unsigned 64-bit integer
+        return UnsignedDecodeInfo{ 8, 0xFFFFFFFFFFFFFFFF };
+    }
+    chipDie();
+}
+
+struct SignedDecodeInfo
+{
+    unsigned byteCount;
+    int64_t minValue;
+    int64_t maxValue;
+
+    template <typename T>
+    static constexpr SignedDecodeInfo From()
+    {
+        return SignedDecodeInfo{
+            sizeof(T),
+            std::numeric_limits<T>::min(),
+            std::numeric_limits<T>::max(),
+        };
+    }
+
+    static constexpr SignedDecodeInfo OfSize(unsigned n)
+    {
+        return SignedDecodeInfo{
+            n,
+            -static_cast<int64_t>(1LL << (8 * n - 1)),
+            (static_cast<int64_t>((1LL << (8 * n - 1))) - 1),
+        };
+    }
+};
+
+constexpr SignedDecodeInfo GetSignedDecodeInfo(EmberAfAttributeType type)
+{
+
+    switch (type)
+    {
+    case ZCL_INT8S_ATTRIBUTE_TYPE: // Unsigned 8-bit integer
+        return SignedDecodeInfo::From<int8_t>();
+    case ZCL_INT16S_ATTRIBUTE_TYPE: // Unsigned 16-bit integer
+        return SignedDecodeInfo::From<int16_t>();
+    case ZCL_INT24S_ATTRIBUTE_TYPE: // Unsigned 24-bit integer
+        return SignedDecodeInfo::OfSize(3);
+    case ZCL_INT32S_ATTRIBUTE_TYPE: // Unsigned 32-bit integer
+        return SignedDecodeInfo::From<int32_t>();
+    case ZCL_INT40S_ATTRIBUTE_TYPE: // Unsigned 40-bit integer
+        return SignedDecodeInfo::OfSize(5);
+    case ZCL_INT48S_ATTRIBUTE_TYPE: // Unsigned 48-bit integer
+        return SignedDecodeInfo::OfSize(6);
+    case ZCL_INT56S_ATTRIBUTE_TYPE: // Unsigned 56-bit integer
+        return SignedDecodeInfo::OfSize(7);
+    case ZCL_INT64S_ATTRIBUTE_TYPE: // Unsigned 64-bit integer
+        return SignedDecodeInfo::From<int64_t>();
+    }
+    chipDie();
+}
+
+static constexpr bool IsOddIntegerSize(unsigned byteCount)
+{
+    // All these conditions seem to result in the same code size:
+    // - (byteCount > 2) && (byteCount != 4) && (byteCount != 8) OR
+    // - (byteCount == 6) || ((byteCount & 0x1) != 0)
+    //
+    // So ended up keeping the "readable" one
+    return (byteCount == 3) || (byteCount == 5) || (byteCount == 6) || (byteCount == 7);
+}
+
+// This is an odd workaround for legacy. Errors SHOULD be always ConstraintError
+// however in practice old ember code returns INVALID_ARGUMENT for odd sized integers
```

Comment:
> Our unit tests validate ASSERT_EQ on the exact error code returned

Well, that's not observable on the wire, right?  The unit test is asserting things that are not actually visible in practice, then, and we should just fix the unit test and whatnot.

What does this code result in on the wire?

Reply:
That would require the test to assume another conversion from CHIP_ERROR to status code.

Doable however outside of the scope of this PR. The issue is that my `check` logic for validation for the datamodel provider  asserts that codegen data model and ember are identical and that assertion is at a `exactly the same CHIP_ERROR is returned` level. The conversion to status-code happens later.

Changing this would require code paths updating. I think at some point we discussed that the action results should be `status codes` and not CHIP_ERROR, which the new API does however updating both new APIs and old ember code is not practical at this point.

Reply:
I think this would be here among other things: https://github.com/project-chip/connectedhomeip/blob/master/src/app/reporting/Read-Checked.cpp#L84

We could make ActionReturnStatus always convert to to Status and compare, however that does not sound ideal if we want to check for bug-for-bug compatibility (and this is indeed a bug I am looking to be compatible with .... I have extra code to reproduce this odd behavior)

Reply:
>  the datamodel provider asserts that codegen data model and ember are identical and that assertion is at a exactly the same CHIP_ERROR is returned level. 

This is the part I don't get: the ember bits _never have that INVALID_ARGUMENT escape_.  The check bits can never see it, as far as I can tell.  So I really don't understand where this is all coming from.

Reply:
I had checks in the mock data model validation (with TODO of the errors not being valid). The error being dependent on odd sized integer check is questionable for sure ...

I updated all tests and made everything constraint errors, will see if CI complains on anything. At least unit tests and python tests pass.

Reply:
It seems the logic was due to the check inside the old https://github.com/project-chip/connectedhomeip/blob/master/src/app/codegen-data-model-provider/CodegenDataModelProvider_Write.cpp#L178C12-L182C123:


```
// This guards against trying to decode something that overlaps nullable, for example
// Nullable<uint8_t>(0xFF) is not representable because 0xFF is the encoding of NULL in ember
// as well as odd-sized integers (e.g. full 32-bit value like 0x11223344 cannot be written
// to a 3-byte odd-sized integger).
VerifyOrReturnError(Traits::CanRepresentValue(isNullable, workingValue.Value()), CHIP_ERROR_INVALID_ARGUMENT);
```

On the surface this looked reasonable (bad input to write returns a "invalid argument") but since only odd sized integers can be outside of the range in this way, it results in odd sized integers returning INVALID_ARGUMENT when constraint seems more uniform.

Now that the code changed, this difference is more obvious and odd. I am unsure if ember checks will hit this, it may not (as we do not often check unhappy paths in integration checks) so I will let CI decide if the change is ok or not.